crates/circuit/src/classical/expr/var.rs

```diff
@@ -0,0 +1,214 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::bit::{ClassicalRegister, ShareableClbit};
+use crate::classical::expr::{ExprKind, PyExpr};
+use crate::classical::types::Type;
+use crate::imports::UUID;
+use pyo3::prelude::*;
+use pyo3::types::{IntoPyDict, PyTuple};
+use pyo3::{intern, IntoPyObjectExt};
+use uuid::Uuid;
+
+/// A classical variable expression.
+///
+/// Note that the type of variant [Var::Bit] is always assumed to be a bool,
+/// so we don't store it.
+#[derive(Clone, Debug, PartialEq, Hash)]
+pub enum Var {
+    Standalone {
+        uuid: u128,
+        name: String,
+        ty: Type,
+    },
+    Bit {
+        bit: ShareableClbit,
+    },
+    Register {
+        register: ClassicalRegister,
+        ty: Type,
+    },
+}
+
+impl<'py> IntoPyObject<'py> for Var {
+    type Target = PyAny;
+    type Output = Bound<'py, PyAny>;
+    type Error = PyErr;
+
+    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+        Ok(Bound::new(py, (PyVar(self), PyExpr(ExprKind::Var)))?.into_any())
+    }
+}
+
+impl<'py> FromPyObject<'py> for Var {
+    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
+        let PyVar(v) = ob.extract()?;
+        Ok(v)
+    }
+}
+
+/// A classical variable.
+///
+/// These variables take two forms: a new-style variable that owns its storage location and has an
+/// associated name; and an old-style variable that wraps a :class:`.Clbit` or
+/// :class:`.ClassicalRegister` instance that is owned by some containing circuit.  In general,
+/// construction of variables for use in programs should use :meth:`Var.new` or
+/// :meth:`.QuantumCircuit.add_var`.
+///
+/// Variables are immutable after construction, so they can be used as dictionary keys."""
+#[pyclass(eq, hash, frozen, extends = PyExpr, name = "Var", module = "qiskit._accelerate.circuit.classical.expr")]
+#[derive(PartialEq, Clone, Debug, Hash)]
+pub struct PyVar(Var);
+
+#[pymethods]
+impl PyVar {
+    #[new]
+    #[pyo3(signature = (var, ty, *, name = None), text_signature="(var, type, *, name=None)")]
+    fn new0(var: &Bound<PyAny>, ty: Type, name: Option<String>) -> PyResult<Py<Self>> {
```

Comment:
As mentioned in [this comment](https://github.com/Qiskit/qiskit/pull/14068/files#r2045333476):
```suggestion
    fn py_new(var: &Bound<PyAny>, ty: Type, name: Option<String>) -> PyResult<Py<Self>> {
```